File: scripts/ie/itemtypes.py
import os
import ruamel.yaml
from .offsets import string_to_id

yaml = ruamel.yaml.YAML(typ="rt")


def get_itemtypes(iesdp_file_formats_dir):
    """
    Returns IESDP item types list {id, desc, value}
    """
    source_file = os.path.join(iesdp_file_formats_dir, "item_types.yml")
    with open(source_file, encoding="utf8") as fhandle:
        items = yaml.load(fhandle)
    itypes = []
    for i in items:
        iid = get_itemtype_id(i)
        if iid == "ITEMTYPE_unknown":
            continue
        itypes.append({"id": iid, "desc": i["type"], "value": i["code"]})
    return itypes


def get_itemtype_id(item):
    """
    Tries to get an id for an item type
    """
    iprefix = "ITEMTYPE_"
    # custom id
    if "id" in item:
        return iprefix + item["id"]
    # no custom id, constructing from description
    iid = string_to_id(item["type"], iprefix)
    return iid


def save_itemtypes_ielib(ielib_structures_dir, itypes):
    """
    Saves item types to IElib structures/item_types.tpp
    """
    dest_file = os.path.join(ielib_structures_dir, "item_types.tpp")
    text = ""
    for itype in itypes:
        text += f"{itype['id']} = {itype['value']}\n"
    with open(dest_file, "w", encoding="utf8") as fhandle:
        print(text, file=fhandle)


def get_itemtypes_isense(itypes):
    """
    Prepares item types list for intellisense
    """
    isense = []
    for itype in itypes:
        iid = itype["id"]
        detail = f"int {iid} = {itype['value']}"
        doc = itype["desc"]
        item = {"name": iid, "detail": detail, "doc": doc}
        isense.append(item)
    return isense

File: scripts/ie-update/src/ie/item-types.ts
/**
 * IESDP item type processing.
 * Extracts item types from IESDP data, generates IElib TPP definitions,
 * and formats item types for IDE completion.
 */

import fs from "node:fs";
import path from "node:path";
import YAML from "yaml";
import { stringToId } from "./offsets.js";
import type { CompletionItem, ItemType, ItemTypeRaw } from "./types.js";
import { validateArray, validateItemTypeRaw } from "./validate.js";

const ITEM_TYPE_PREFIX = "ITEM_TYPE_";

/**
 * Loads item types from IESDP file_formats/item_types.yml.
 * Generates IDs from descriptions, skipping unknown types.
 */
export function getItemTypes(iesdpFileFormatsDir: string): readonly ItemType[] {
    const sourceFile = path.join(iesdpFileFormatsDir, "item_types.yml");
    const content = fs.readFileSync(sourceFile, "utf8");
    const items: readonly ItemTypeRaw[] = validateArray(
        YAML.parse(content),
        validateItemTypeRaw,
        sourceFile,
    );

    return items.reduce<readonly ItemType[]>((acc, item) => {
        const iid = getItemTypeId(item);
        if (iid === "ITEM_TYPE_unknown") {
            return acc;
        }
        if (Number.isNaN(Number(item.code))) {
            throw new Error(`Invalid item type code '${item.code}' for '${item.type}' in ${sourceFile}`);
        }
        return [...acc, { id: iid, desc: item.type, value: item.code }];
    }, []);
}

/**
 * Gets or generates an ID for an item type.
 * Uses custom 'id' field if present, otherwise constructs from the type description.
 */
function getItemTypeId(item: ItemTypeRaw): string {
    if (item.id !== undefined) {
        return ITEM_TYPE_PREFIX + item.id;
    }
    return stringToId(item.type, ITEM_TYPE_PREFIX);
}

/**
 * Saves item types as IElib TPP defines to structures/item_types.tpp.
 */
export function saveItemTypesIelib(ielibStructuresDir: string, itypes: readonly ItemType[]): void {
    const destFile = path.join(ielibStructuresDir, "item_types.tpp");
    const text = itypes.map((itype) => `${itype.id} = ${itype.value}`).join("\n") + "\n";
    fs.writeFileSync(destFile, text + "\n", "utf8");
}

/**
 * Formats item types for IDE intellisense completion items.
 */
export function getItemTypesIsense(itypes: readonly ItemType[]): readonly CompletionItem[] {
    return itypes.map((itype) => ({
        name: itype.id,
        detail: `int ${itype.id} = ${itype.value}`,
        doc: itype.desc,
    }));
}
